### PR TITLE
Implement API access control when fetching profiles

### PR DESCRIPTION
**Pull request by sbathgate**
Resolves issue #49. To achieve issue #73 Create a dashboard where the user can view their existing projects, an access control feature is necessary so a regular user can only access their profiles, while an administrator can access all profiles.

This PR implements an access control capability when requesting access to profiles:

- [X] Update fetchAllProjectProfiles() to include access control
- [X] Update fetchProjectProfile() to include access control
- [X] implement findProfilesByUserID() function
- [x] profile unit testing
- [x] update API docs

### CODE REVIEW

File: api/src/controllers/profile.ts [hunk 1]
```diff
@@ -67,15 +73,20 @@ export const fetchAllProjectProfiles = async (req: Request, res: Response): Prom
...
+  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
 ): Promise<void> => {
   const { ProfileModel } = dm;
   const { profileId } = params;
 
   try {
     const results = await ProfileModel.findById(Number(profileId));
 
-    res.status(200).json(results);
+    if (user.id === results.userId || user.roles.includes('administrator')) {
```
Comment (jleach) [inline api/src/controllers/profile.ts:84]: I think I'd put "administrator" into `constants.ts`.

Comment (jleach) [inline api/src/controllers/profile.ts:84]: I generally prefer early returns to if/else where possible; I think it makes the code more readable and less error prone. For example, one could:
```
if (user.id !== results.userId || !user.roles.includes('administrator')) {
   throw errorWithCode('Unauthorized Access', 401);
}
```

And skip the else.

File: api/src/controllers/profile.ts [hunk 2]
```diff
@@ -67,15 +73,20 @@ export const fetchAllProjectProfiles = async (req: Request, res: Response): Prom
...
+      res.status(200).json(results);
+    } else {
+      throw errorWithCode('Unauthorized Access', 401);
```
Comment (jleach) [inline api/src/controllers/profile.ts:87]: I don't think this error will make it out. It'll get caught by the catch below and replaced with `throw errorWithCode(message, 500)`; You probably need to check and re-throw.
```
if (err.code) {
  throw err
}
```

**Review by jleach — CHANGES_REQUESTED** [review]
Looking good. I think as part of this ticket you should do all the other endpoints. For example,  `updateProfileNamespace`. Lets get all the CRUD ops done in this one ticket.

File: api/src/controllers/profile.ts [hunk 3]
```diff
@@ -51,11 +51,17 @@ export const uniqueNamespacePrefix = async (): Promise<string | undefined> => {
...
+  { user }: { user: AuthenticatedUser }, res: Response): Promise<void> => {
   const { ProfileModel } = dm;
 
   try {
-    const results = await ProfileModel.findAll();
+    let results;
+    if (user.roles.includes('administrator')) {
+      results = await ProfileModel.findAll();
+    } else {
+      results = await ProfileModel.findProfilesByUserID(user.id);
```
[the remark this answers is not in the document]
Comment (sbathgate) [reply api/src/controllers/profile.ts]: Good catch. Adjusted to match camelCase use throughout site. Additionally caught other inconsistencies which I updated accordingly.

**Review by jleach — APPROVED** [review]

**Pull request merged**